airflow/io/fs/__init__.py

```diff
@@ -0,0 +1,966 @@
+
+        return new_path
+
+
+MOUNTS: dict[str:Mount] = {}
+
+
+def get_mount(path: str) -> Mount:
+    """
+    Get the mount point for a given path.
+
+    :param path: the path to get the mount point for
+    :type path: str
+    :return: the mount point
+    :rtype: str
+    """
+    mount_point = None
+    mount_points = sorted(MOUNTS.keys(), key=len, reverse=True)
+
+    for prefix in mount_points:
+        if os.path.commonprefix([prefix, path]) == prefix:
+            mount_point = prefix
+
+    if mount_point is None:
+        raise ValueError(f"No mount point found for path: {path}")
+
+    return MOUNTS.get(mount_point)
+
+
+def _replace_mount_point(path: str) -> str:
+    """
+    Replace the mount point in a path with the original source.
+
+    :param path: the path to replace the mount point in
+    :type path: str
+    :return: the path with the mount point replaced
+    :rtype: str
+    """
+    mnt = get_mount(path)
+
+    return path.replace(mnt.mount_point, mnt.source)
+
+
+def _rewrite_path(path: str, mnt: Mount) -> str:
+    """
+    Rewrite a path to include the mount point and remove the original source.
+
+    :param path: the path to rewrite
+    :type path: str
+    :param mnt: the mount point to include in the path
+    :type mnt: Mount
+    :return: the rewritten path
+    :rtype: str
+    """
+    return os.path.join(mnt.mount_point, path.replace(mnt.source, "").lstrip(os.sep))
+
+
+def _rewrite_info(info: dict, mnt: Mount) -> dict:
+    """
+    Rewrite the path in a file info dict to include the mount point and remove the original source.
+
+    :param info: the file info dict to rewrite
+    :type info: dict
+    :param mnt: the mount point to include in the path
+    :type mnt: Mount
+    :return: the rewritten file info dict
+    :rtype: dict
+    """
+    info["original_name"] = info["name"]
+    info["name"] = _rewrite_path(info["name"], mnt)
+
+    return info
+
+
+def mount(
+    source: str,
+    mount_point: str | None = None,
+    conn_id: str | None = None,
+    encryption_type: str | None = "",
+    fs_type: AbstractFileSystem | None = None,
+    remount: bool = False,
+) -> Mount:
+    """
+    Mount a filesystem or object storage to a mount point.
+
+    :param source: the source path to mount
+    :type source: str
+    :param mount_point: the target mount point
+    :type mount_point: str
+    :param conn_id: the connection to use to connect to the filesystem
+    :type conn_id: str
+    :param encryption_type: the encryption type to use to connect to the filesystem
+    :type encryption_type: str
+    :param fs_type: the filesystem type to use to connect to the filesystem
+    :type fs_type: AbstractFileSystem
+    :param remount: whether to remount the filesystem if it is already mounted
+    :type remount: bool
+    """
+    if not remount and mount_point and mount_point in MOUNTS:
+        raise ValueError(f"Mount point {mount_point} already mounted")
```

Comment:
I wonder if we should return the existing mount if it is created with exactly the same arguments…?